Added tf.math.zeta to tf frontend

akshatvishu opened the pull request:
Close #13313

akshatvishu commented:
Several tests have failed during the review of this PR. Specifically, the zeta function is failing tests on the NumPy, JAX, and Paddle backends.

- For the NumPy backend, it seems like there is an issue with trying to assign to a scalar value. The error message indicates that a 'numpy.float64' object does not support item assignment. We need to modify the NumPy backend implementation to use a valid index when assigning values.
- 
- For the JAX backend, it looks like there is an issue with using a boolean scalar index. The error message indicates that JAX arrays do not support boolean scalar indices. We should modify the JAX implementation to use a different type of index, such as an integer index.
- 
- Finally, for the Paddle backend, the zeta function is not yet implemented. We need to add an implementation for it in ivy/functional/backends/paddle/math.py. 
- 

 I  welcome any suggestions or insights on how to address these issues and ensure that the tests pass successfully. Thank you in advance for your help

AnnaTz commented on ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py:
```diff
@@ -1964,4 +1964,37 @@ def test_tensorflow_sinh(
+# zeta
+@handle_frontend_test(
+    fn_tree="tensorflow.math.zeta",
+    dtype_and_x=helpers.dtype_and_values(
```
You'll need to add an argument here for the alias raw_ops function. Here you can read the documentation on how we are testing aliases https://lets-unify.ai/ivy/overview/deep_dive/ivy_frontends_tests.html#alias-functions.

AnnaTz commented on ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py:
```diff
@@ -1964,4 +1964,37 @@ def test_tensorflow_sinh(
+        available_dtypes=[
+            "float32",
+            "float64",
+        ],
```
As long as we have defined the supported dtypes for the functions using the appropriate decorators, we don't need to define a limited set of available_dtypes in the test. We should be testing for all valid dtypes.

akshatvishu commented:
> I think the numpy and jax errors are caused by providing 0-dimensional arrays to the function.

Is there anything I could to change that behavior?

> As long as we have defined the supported dtypes for the functions using the appropriate decorators, we don't need to define a limited set of available_dtypes in the test. We should be testing for all valid dtype

Then should I change the `dtype` to  `numeric`  or  to `float` because `tf.math.zeta` considers only `float32` and `float64`  as valid input `dtypes`. 

> You'll need to add an argument here for the alias raw_ops function. Here you can read the documentation on how we are testing aliases https://lets-unify.ai/ivy/overview/deep_dive/ivy_frontends_tests.html#alias-functions.

Sure! 

> You're not supposed to be making changes to files that are not related to the functions you're implementing. Please revert these.

Sorry, my bad!

akshatvishu commented:
Hey @AnnaTz , I made changes to `test_function` to add `min_num_dims=1` and also added the alias. Currently, dealing with `Assertion Error` for `jax` and `numpy` and type error for `tensorflow` and `torch`.

```bash
FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.numpy-False-False] - AssertionError:  the results from backend numpy and ground truth framework tensorflow do not match
FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.jax-False-False] - AssertionError:  the results from backend jax and ground truth framework tensorflow do not match
FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.tensorflow-False-False] - TypeError: _wraped_fn() takes 0 positional arguments but 1 was given
FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.torch-False-False] - TypeError: _wraped_fn() takes 0 positional arguments but 1 was given
FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.paddle-False-False] - ivy.utils.exceptions.IvyNotImplementedException
```

The above message for `TypeError`  suggests that _wrapped_fn() takes 0 positional arguments but 1 was given, which is usually an indication that the function is expecting to be called with some argument(s). In this case, since the test is setting `test_with_out=st.just(False)`, which means that the out argument is not being used, it could be the cause of the error.

Maybe, I should try modifying the test to set `test_with_out=st.just(True)` instead and provide a valid `out` argument.  Again on any lead on where to look-at to correct the above will be highly appreciated!

AnnaTz commented:
> Hey @AnnaTz , I made changes to `test_function` to add `min_num_dims=1` and also added the alias. Currently, dealing with `Assertion Error` for `jax` and `numpy` and type error for `tensorflow` and `torch`.
> 
> ```shell
> FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.numpy-False-False] - AssertionError:  the results from backend numpy and ground truth framework tensorflow do not match
> FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.jax-False-False] - AssertionError:  the results from backend jax and ground truth framework tensorflow do not match
> FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.tensorflow-False-False] - TypeError: _wraped_fn() takes 0 positional arguments but 1 was given
> FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.torch-False-False] - TypeError: _wraped_fn() takes 0 positional arguments but 1 was given
> FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.paddle-False-False] - ivy.utils.exceptions.IvyNotImplementedException
> ```
> 
> The above message for `TypeError` suggests that _wrapped_fn() takes 0 positional arguments but 1 was given, which is usually an indication that the function is expecting to be called with some argument(s). In this case, since the test is setting `test_with_out=st.just(False)`, which means that the `out` argument is not being used, it could be the cause of the error.
> 
> Maybe, I should try modifying the test to set `test_with_out=st.just(True)` instead and provide a valid `out` argument. Again on any lead on where to look-at to correct the above will be highly appreciated!

My guess is this error is related to the way you have defined the alias function. What if you remove the `with_supported_dtypes` decorator there?

akshatvishu commented:
> > Hey @AnnaTz , I made changes to `test_function` to add `min_num_dims=1` and also added the alias. Currently, dealing with `Assertion Error` for `jax` and `numpy` and type error for `tensorflow` and `torch`.
> > ```shell
> > FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.numpy-False-False] - AssertionError:  the results from backend numpy and ground truth framework tensorflow do not match
> > FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.jax-False-False] - AssertionError:  the results from backend jax and ground truth framework tensorflow do not match
> > FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.tensorflow-False-False] - TypeError: _wraped_fn() takes 0 positional arguments but 1 was given
> > FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.torch-False-False] - TypeError: _wraped_fn() takes 0 positional arguments but 1 was given
> > FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.paddle-False-False] - ivy.utils.exceptions.IvyNotImplementedException
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The above message for `TypeError` suggests that _wrapped_fn() takes 0 positional arguments but 1 was given, which is usually an indication that the function is expecting to be called with some argument(s). In this case, since the test is setting `test_with_out=st.just(False)`, which means that the `out` argument is not being used, it could be the cause of the error.
> > Maybe, I should try modifying the test to set `test_with_out=st.just(True)` instead and provide a valid `out` argument. Again on any lead on where to look-at to correct the above will be highly appreciated!
> 
> My guess is this error is related to the way you have defined the alias function. What if you remove the `with_supported_dtypes` decorator there?

Hey @AnnaTz , thanks for the suggestions, I tried these scenario to pin-point just what might be causing the error:

-  Removed `with_supported_dtypes` decorator from alias and run the test : 
 => This resulted in the same  `TypeError` : `_wrapped_fn()` takes 0 positional arguments but 1 was given.
-  Removed the alias from test_function and then run it:
 =>This causes `AssertionError` for all the backend except Paddle because `paddle` dosen't have a `zeta` implementation 
 
 Now, looking at ` _wraped_fn()` function inside the `func_wrapper.py` file, which only accepts **keyword arguments** currently, I am thinking that will changing it to accept **positional arguments** as well can remove the error!

AnnaTz commented:
> > > Hey @AnnaTz , I made changes to `test_function` to add `min_num_dims=1` and also added the alias. Currently, dealing with `Assertion Error` for `jax` and `numpy` and type error for `tensorflow` and `torch`.
> > > ```shell
> > > FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.numpy-False-False] - AssertionError:  the results from backend numpy and ground truth framework tensorflow do not match
> > > FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.jax-False-False] - AssertionError:  the results from backend jax and ground truth framework tensorflow do not match
> > > FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.tensorflow-False-False] - TypeError: _wraped_fn() takes 0 positional arguments but 1 was given
> > > FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.torch-False-False] - TypeError: _wraped_fn() takes 0 positional arguments but 1 was given
> > > FAILED ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py::test_tensorflow_zeta[cpu-ivy.functional.backends.paddle-False-False] - ivy.utils.exceptions.IvyNotImplementedException
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The above message for `TypeError` suggests that _wrapped_fn() takes 0 positional arguments but 1 was given, which is usually an indication that the function is expecting to be called with some argument(s). In this case, since the test is setting `test_with_out=st.just(False)`, which means that the `out` argument is not being used, it could be the cause of the error.
> > > Maybe, I should try modifying the test to set `test_with_out=st.just(True)` instead and provide a valid `out` argument. Again on any lead on where to look-at to correct the above will be highly appreciated!
> > 
> > 
> > My guess is this error is related to the way you have defined the alias function. What if you remove the `with_supported_dtypes` decorator there?
> 
> Hey @AnnaTz , thanks for the suggestions, I tried these scenarios to pin-point just what might be causing the error:
> 
> * Removed `with_supported_dtypes` decorator from alias and run the test :
>   => This resulted in the same  `TypeError` : `_wrapped_fn()` takes 0 positional arguments but 1 was given.
> * Removed the alias from test_function and then run it:
>   =>This causes `AssertionError` for all the backend except Paddle because `paddle` dosen't have a `zeta` implementation
> 
> Now, looking at ` _wraped_fn()` function inside the `func_wrapper.py` file, which only accepts **keyword arguments** currently, I am thinking that will changing it to accept **positional arguments** as well can remove the error!

You can ignore the paddle assertion error, so what remains to be resolved is the alias function.

akshatvishu commented:
> 
> You can ignore the paddle assertion error, so what remains to be resolved is the alias function.

Yep, I hope this was the correct way to write the alias :
```python
# zeta
@handle_frontend_test(
    fn_tree="tensorflow.math.zeta",
    aliases=["tensorflow.raw_ops.Zeta"],
    dtype_and_x=helpers.dtype_and_values(
        available_dtypes=helpers.get_dtypes("valid"),
        min_num_dims=1,
        num_arrays=2,
        shared_dtype=True,
    ),
    test_with_out=st.just(False),
)
def test_tensorflow_zeta(
    *,
    dtype_and_x,
    frontend,
    test_flags,
    fn_tree,
    on_device,
):
    input_dtype, x = dtype_and_x
    helpers.test_frontend_function(
        input_dtypes=input_dtype,
        frontend=frontend,
        test_flags=test_flags,
        fn_tree=fn_tree,
        on_device=on_device,
        x=x[0],
        q=x[1],
    )

```

AnnaTz commented:
> > You can ignore the paddle assertion error, so what remains to be resolved is the alias function.
> 
> Yep, I hope this was the correct way to write the alias :
> 
> ```python
> # zeta
> @handle_frontend_test(
>     fn_tree="tensorflow.math.zeta",
>     aliases=["tensorflow.raw_ops.Zeta"],
>     dtype_and_x=helpers.dtype_and_values(
>         available_dtypes=helpers.get_dtypes("valid"),
>         min_num_dims=1,
>         num_arrays=2,
>         shared_dtype=True,
>     ),
>     test_with_out=st.just(False),
> )
> def test_tensorflow_zeta(
>     *,
>     dtype_and_x,
>     frontend,
>     test_flags,
>     fn_tree,
>     on_device,
> ):
>     input_dtype, x = dtype_and_x
>     helpers.test_frontend_function(
>         input_dtypes=input_dtype,
>         frontend=frontend,
>         test_flags=test_flags,
>         fn_tree=fn_tree,
>         on_device=on_device,
>         x=x[0],
>         q=x[1],
>     )
> ```

The existing tensorflow.raw_ops functions have separate test functions, not sure why. You could just write the test manually instead of defining the `aliases` argument.

akshatvishu commented:
Hi !@AnnaTz , I added the `test_fn` for `Zeta`  in `test_raw_ops.py` but here I changed the `available_dtype` in `test_fn` to `available_dtype= ["float32", "float64"]` because adding `with_support_dtypes` in the `raw_ops.py` for `Zeta` was throwing error! 

Also, If my local branch is behind the remote branch, should doing `git pull` -> `git add .`  ->` git commit` the way to go or should I handle it differently ?

AnnaTz commented:
> Also, If my local branch is behind the remote branch, should doing git pull -> git add . -> git commit the way to go or should I handle it differently ?

I do `git stash` -> `git pull` -> `git stash apply`.

akshatvishu commented:
> You'll first need to sync your master with our master on github. Then, on your local master do `git pull`, checkout to your branch and do `git stash` -> `git merge master` -> `git stash apply`.

I did that but these three files are giving me error:
```python
Unmerged paths:
  (use "git add/rm <file>..." as appropriate to mark resolution)
        deleted by us:   .devcontainer/build/devcontainer.json
        both modified:   ivy/functional/frontends/tensorflow/math.py
        both modified:   ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py


vscode ➜ /workspaces/ivy (adding_zeta_ivy_tf_frontend) $ git add ivy/functional/frontends/tensorflow/math.py 
error: insufficient permission for adding an object to repository database .git/objects
error: ivy/functional/frontends/tensorflow/math.py: failed to insert into database
error: unable to index file 'ivy/functional/frontends/tensorflow/math.py'
fatal: updating files failed

```

AnnaTz commented:
> > You'll first need to sync your master with our master on github. Then, on your local master do `git pull`, checkout to your branch and do `git stash` -> `git merge master` -> `git stash apply`.
> 
> Hey @AnnaTz , I tried to do as you instructed but these three files are giving me error:
> 
> ```python
> Unmerged paths:
>   (use "git add/rm <file>..." as appropriate to mark resolution)
>         deleted by us:   .devcontainer/build/devcontainer.json
>         both modified:   ivy/functional/frontends/tensorflow/math.py
>         both modified:   ivy_tests/test_ivy/test_frontends/test_tensorflow/test_math.py
> 
> 
> vscode ➜ /workspaces/ivy (adding_zeta_ivy_tf_frontend) $ git add ivy/functional/frontends/tensorflow/math.py 
> error: insufficient permission for adding an object to repository database .git/objects
> error: ivy/functional/frontends/tensorflow/math.py: failed to insert into database
> error: unable to index file 'ivy/functional/frontends/tensorflow/math.py'
> fatal: updating files failed
> ```
> 
> when try to `git add` or `git rm` .devcontainer:
> 
> ```
> vscode ➜ /workspaces/ivy (adding_zeta_ivy_tf_frontend) $ git add .devcontainer/build/devcontainer.json
> error: insufficient permission for adding an object to repository database .git/objects
> error: .devcontainer/build/devcontainer.json: failed to insert into database
> error: unable to index file '.devcontainer/build/devcontainer.json'
> fatal: updating files failed
> ```

Are you on VScode? I'm afraid I'm not sure how to help you with this. Please ask on our discord server.

akshatvishu commented:
> Are you on VScode? I'm afraid I'm not sure how to help you with this. Please ask on our discord server.

Sure thing @AnnaTz  , I asked it on #vscode channel but yet to get a response! In the worst-case  scenario should I open the new PR for the same issue ? (Will wait for a day to see it I can get a response on discord)

akshatvishu commented:
Hey @AnnaTz, I regret to inform you that I will have to close this pull request. Unfortunately, I have been unable to find a solution to the error message presented below:
```bash
error: unable to unlink old '.devcontainer/build/devcontainer.json': Permission denied
``` 
I will be opening a new pull request in order to address the same issue.

akshatvishu commented:
Hey @AnnaTz ,  I apologize for any inconvenience caused by my repeated pinging. I would like to inform you that I have opened a new pull request for the same issue, which can be found at this link: [new PR](https://github.com/unifyai/ivy/pull/14125). I would greatly appreciate it if you could kindly review the new pull request when you have a chance. Thank you very much for your time and assistance.
